### ISSUE TITLE
How to handle cores in kubernetes

### ISSUE THREAD

**Issue by diazGT94** [aws/aws-neuron-sdk#392]
Hello,

I have an EKS cluster that contains 1 node with 1inf instance. Currently I'm able to deploy a Docker Image on that node and it's working. However I would like to know what I need to do to deploy 2 different containers in parallel on the same node, in a way 1 container uses 2 cores and the other uses the other 2 cores. In my YAML I've the following configuration:

``` resources:
            limits:
              aws.amazon.com/neuron: '1'
              cpu: '4'
              memory: 4Gi
            requests:
              cpu: '2'
              memory: 1Gi
```
But when I tried to run two containers in parallel I get the following error:  `Unschedulable: 0/1 nodes are available: 1 Insufficient aws.amazon.com/neuron.`

Thanks for the help.

**Comment by aws-owinop** [comment]
Hi @diazGT94,
The plug-in only supports device allocation and doesn't support core allocations today. We'll add a feature request for supporting core allocations in the future. You can follow our documentation [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-deploy/tutorials/tutorial-k8s.html#step-1-deploy-the-neuron-device-plugin) on how to specify devices for the container in kubernetes.
Note that Neuron device logical IDs must be contiguous. If a pod requesting multiple Neuron devices is scheduled on an inf1.6xlarge or inf1.24xlarge instance type (which have more than one Neuron device), that pod will fail to start if the Kubernetes scheduler selects non-contiguous device IDs. Here is the relevant EKS documentation https://docs.aws.amazon.com/eks/latest/userguide/inferentia-support.html

**Comment by alejoGT1202** [comment]
Hello @aws-owinop do you have any idea if this function will be implemented in the first semester of this year?

**Comment by aws-owinop** [comment]
Hi @alejoGT1202, [Kubernetes](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) device plugin does not support device sharing across containers.  We implemented the Neuron k8 scheduler to expose NeuronDevices (inferentia) to k8, not the NeuronCores.
Do you mind elaborating a little more on your need for supporting core allocations at the container level instead of device?  It will help us understand and plan for any future changes.

**Comment by alejoGT1202** [comment]
I have a process that requires to run multiple models in parallel before triggering the next step in my pipeline, I know that this can be done using one instance that has more than one NeuronDevice, but I noticed that in terms of processing time there is not a significant difference between loading a model in the 4 cores of a device against loading a model in one core of a device. If a load one model in each core of the device I can process more data in a similar time that loading the model in the full device, which is the kind of solution I'm trying to do right now, trying to optimize the usage of the cores.

**Comment by aws-owinop** [comment]
Thanks for the additional info. Is this issue related to [393](https://github.com/aws/aws-neuron-sdk/issues/393)

**Comment by alejoGT1202** [comment]
Yes it's related to the that issue.

**Comment by zoltan-fedor** [comment]
I hit the same issue - as I assume most people who will want to use Inferentia from a Kubernetes cluster - with the exception of Inferentia-dedicated clusters.

Basically if I am reading this right, those nice blog posts from AWS about how great the Inferentia chip is on EKS forgot to mention, that that really only works economically if you have an EKS cluster ONLY containing Inferentia nodes, otherwise you will hit this little issue.

I guess most people - just like me - will have an EKS cluster where SOME of the nodes will have Inferentia cores and they will want to assign containers requiring access to Inferentia cores to those nodes. That currently can be done thanks to this plugin, but only a SINGLE CONTAINER can be assigned - which makes the use of Inferentia nodes commercially NOT viable. Which is pretty much the point over GPU nodes. Due to this it seems GPU nodes are actually (a lot) more cost-effective - until this issue isn't fixed.

@micwade-aws, Thanks for labeling this issue as 'enhancement'!
Would it be possible to handle this with priority - as clearly it just kills the commercial viability of the Inferentia chips usage on Kubernetes (with the exceptions of dedicated Kubernetes clusters where no scheduling is required as all nodes have the Inferentia chip)

 Thanks

**Comment by micwade-aws** [comment]
@zoltan-fedor @diazGT94 @alejoGT1202  - would you mind connecting with us over email at [aws-neuron-support@amazon.com](mailto:aws-neuron-support@amazon.com)?  We implemented a prototype that does what you're asking, but have not committed it to a release for other reasons.  Would be great to give you the prototype as an early preview and unblock your workflow.

**Comment by zoltan-fedor** [comment]
@micwade-aws, absolutely. I will email aws-neuron-support@amazon.com, as I would be happy to try out a prototype.
